File: 20170629-ts-gilded_rose/app/gilded-rose.ts
import { Item } from './item';

export class GildedRose {
    items: Array<Item>;

    constructor(items = []) {
        this.items = items;
    }

    updateQuality() {
        return this.items.map(item => this.updateQualityOf(item));
    }

    private updateQualityOf(item: Item): Item {
        if (item.name == 'Sulfuras, Hand of Ragnaros') {
            return item;
        }

        if (item.name == 'Aged Brie') {
            this.increaseQuality(item);

            this.decreaseSellIn(item);

            if (item.sellIn < 0) {
                this.increaseQuality(item);
            }
        } else if (item.name == 'Backstage passes to a TAFKAL80ETC concert') {

            this.increaseQuality(item);
            if (item.sellIn <= 10) {
                this.increaseQuality(item);
            }
            if (item.sellIn <= 5) {
                this.increaseQuality(item);
            }

            this.decreaseSellIn(item);

            if (item.sellIn < 0) {
                this.dropQuality(item);
            }
        } else {
            this.decreaseQuality(item);

            this.decreaseSellIn(item);

            if (item.sellIn < 0) {
                this.decreaseQuality(item);
            }
        }

        return item;
    }

    private decreaseSellIn(item: Item) {
        item.sellIn = item.sellIn - 1;
    }

    private dropQuality(item: Item) {
        item.quality = item.quality - item.quality;
    }

    private decreaseQuality(item: Item) {
        if (item.quality > 0) {
            item.quality = item.quality - 1;
        }
    }

    private increaseQuality(item: Item) {
        if (item.quality < 50) {
            item.quality = item.quality + 1;
        }
    }
}

File: 20190702-java-gilded-rose/src/main/java/gildedrose/GildedRose.java
package gildedrose;

class GildedRose {
    Item[] items;

    public GildedRose(Item[] items) {
        this.items = items;
    }

    public void updateQuality() {
        for (Item item : items) {
            updateQualityOf(item);
        }
    }

    private void updateQualityOf(Item item) {
        switch (item.name) {
            case "Aged Brie":
                updateQualityAgedBrie(item);
                break;
            case "Backstage passes to a TAFKAL80ETC concert":
                updateQualityBackstagePasses(item);
                break;
            case "Sulfuras, Hand of Ragnaros":
                updateQualitySulfuras(item);
                break;
            default:
                updateQualityRegularItem(item);
                break;
        }
    }

    private void updateQualityAgedBrie(Item item) {
        if (item.quality < 50) {
            increaseQuality(item);
        }

        decreaseSellIn(item);

        if (item.sellIn < 0) {
            if (item.quality < 50) {
                increaseQuality(item);
            }
        }
    }

    private void updateQualityBackstagePasses(Item item) {
        if (item.quality < 50) {
            increaseQuality(item);

            if (item.sellIn < 11) {
                if (item.quality < 50) {
                    increaseQuality(item);
                }
            }

            if (item.sellIn < 6) {
                if (item.quality < 50) {
                    increaseQuality(item);
                }
            }
        }

        decreaseSellIn(item);

        if (item.sellIn < 0) {
            dropQuality(item);
        }
    }

    private void updateQualitySulfuras(Item item) {
    }

    private void updateQualityRegularItem(Item item) {
        if (item.quality > 0) {
            decreaseQuality(item);
        }

        decreaseSellIn(item);

        if (item.sellIn < 0) {
            if (item.quality > 0) {
                decreaseQuality(item);
            }
        }
    }

    private void decreaseQuality(Item item) {
        item.quality = item.quality - 1;
    }

    private void increaseQuality(Item item) {
        item.quality++;
    }

    private void decreaseSellIn(Item item) {
        item.sellIn--;
    }

    private void dropQuality(Item item) {
        item.quality = 0;
    }
}